Add BMI160 support

flaviut opened the pull request:
# What does this implement/fix?

Adds support for [the Bosch BMI160 IMU](https://www.bosch-sensortec.com/products/motion-sensors/imus/bmi160/)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/3086

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
sensor:
  - platform: bmi160
    address: 0x68
    update_interval: 60s
    acceleration_x:
      name: "BMI160 Accel X"
    acceleration_y:
      name: "BMI160 Accel Y"
    acceleration_z:
      name: "BMI160 Accel z"
    gyroscope_x:
      name: "BMI160 Gyro X"
    gyroscope_y:
      name: "BMI160 Gyro Y"
    gyroscope_z:
      name: "BMI160 Gyro z"
    temperature:
      name: "BMI160 Temperature"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

flaviut committed:
Add BMI160 support

flaviut commented on esphome/components/bmi160/bmi160.cpp:
```diff
@@ -0,0 +1,244 @@
+    return;
+  }
+  ESP_LOGV(TAG, "  Waiting for gyroscope to wake up...");
+  delay(81);
```
Unfortunately I don't see a better way to do this--I could do a loop checking the I2C register if it's done yet, but that's a lot of work to maybe save 30ms (typical duration of this step is 50ms, max is 81ms)

clydebarrow replied:
You could extract the succeeding code into another function and use `set_timeout()` to call it later.

You'd need a flag for setup complete which you would test in `update()`.

flaviut replied:
Good call! Thank you and done.

flaviut committed:
BMI160: use set_timeout for delay

flaviut committed:
Add support for old compilers

Fix "warning: missing terminating ' character"

clydebarrow commented on esphome/components/bmi160/bmi160.cpp:
```diff
@@ -0,0 +1,260 @@
+      }
+      ESP_LOGV(TAG, "  Waiting for gyroscope to wake up...");
+      // wait between 51 & 81ms, doing 81 to be safe
+      this->set_timeout(81, [this]() { this->internal_setup_(2); });
```
If 81ms is the maximum per the datasheet, I'd allow a little extra in case the timeout call happens early. There is no guarantee on the accuracy of the timeout, and while I'm fairly certain it will not happen early, an extra few ms delay will do no harm. I'd round it up to 100ms.

clydebarrow commented on esphome/components/bmi160/bmi160.cpp:
```diff
@@ -0,0 +1,260 @@
+    this->status_set_warning();
+    return;
+  }
+  int16_t int_temperature = *reinterpret_cast<int16_t *>(&raw_temperature);
```
This relies on the executing architecture being little-endian. Better to explicitly shift and add the bytes, that way it's portable across architecures.

clydebarrow commented on esphome/components/bmi160/sensor.py:
```diff
@@ -0,0 +1,84 @@
+
+DEPENDENCIES = ["i2c"]
+
+CONF_ACCEL_X = "accel_x"
```
There are already similar config constants defined in esphome.const:
```
CONF_ACCELERATION_X = "acceleration_x"
CONF_ACCELERATION_Y = "acceleration_y"
CONF_ACCELERATION_Z = "acceleration_z"

```
I know that other gyro components define their own, but if they're in the system files I would think it better to use those. There are currently no GYRO versions in const.py - maybe you could add those as part of the PR.

clydebarrow commented on esphome/components/bmi160/sensor.py:
```diff
@@ -0,0 +1,84 @@
+
+accel_schema = sensor.sensor_schema(
+    unit_of_measurement=UNIT_METER_PER_SECOND_SQUARED,
+    icon=ICON_BRIEFCASE_DOWNLOAD,
```
There are icons defined in const.py for acceleration in three axes.

```
ICON_ACCELERATION_X = "mdi:axis-x-arrow"
ICON_ACCELERATION_Y = "mdi:axis-y-arrow"
ICON_ACCELERATION_Z = "mdi:axis-z-arrow"
```

Or just `ICON_ACCELERATION` if you don't care about differentiating them.

flaviut committed:
Increase power-on delay to be more conservative

flaviut committed:
Add helper for reading little-endian data over i2c

flaviut committed:
Replace configuration names with globals

Note: for testing with external components, you will need to comment out the import & define your own CONF_GYROSCOPE_X, etc, in this file

flaviut committed:
Improve icons

flaviut commented:
Good call on the icons & configs! Fixed & tested everything.

flaviut committed:
Fix tests & lint

flaviut commented:
Hi I want to check in on this. I understand if it won't be merged, but I would like to at least have the icon changes merged so I can keep this custom integration in an upstreamable state.
